heron/common/src/python/argparser.py

```diff
@@ -0,0 +1,196 @@
+    HeronRCArgumentParser.initializeFromRC(rcfile)
+    super(HeronRCArgumentParser, self).__init__(*args, **kwargs)
+
+  @classmethod
+  def remove_comments(cls, string):
+    pattern = r"(\#.*$)"
+    # first group captures quoted strings (double or single)
+    # second group captures comments (//single-line or /* multi-line */)
+    regex = re.compile(pattern, re.MULTILINE | re.DOTALL)
+
+    def _replacer(match):
+      # if the 2nd group (capturing comments) is not None,
+      # it means we have captured a non-quoted (real) comment string.
+      if match.group(1) is not None:
+        return ""  # so we will return empty to remove the comment
+      else:  # otherwise, we will return the 1st group
+        return match.group(1)  # captured quoted-string
+    return regex.sub(_replacer, string)
+
+  @classmethod
+  def getAndRemoveKey(cls, dictionary, key):
+    val = None
+    if key in dictionary:
+      val = dictionary[key]
+      del dictionary[key]
+    return val
+
+  # initialize the command map from heron rc file in the parser,
+  # that can be later used for command substitution during parse_args phase
+  # patterns
+
+  @classmethod
+  def initializeFromRC(cls, rcfile):
+    if len(cls.cmdmap) > 0:
+      return
+    effective_rc = (rcfile, HERON_RC)[rcfile is None]
+    Log.debug('Effective RC file is %s', effective_rc)
+    if os.path.exists(effective_rc):
+      with open(effective_rc) as f:
+        cls.cmdmap['*']['*'] = collections.defaultdict(dict)
+        cls.cmdmap['*']['*']['*'] = ''
+        for line in f:
+          m = heron_command_pattern.match(line)
+          app, value, command, env = '', '', '', ''
+          if m is not None:
+            value = cls.remove_comments(m.group(4).rstrip(os.linesep))
+            app = (m.group(1), '')[m.group(1) is None or m.group(1) == '']
+            command = (m.group(2), '')[m.group(2) is None or m.group(1) == '']
+            env = (m.group(3), '')[m.group(3) is None or m.group(2) == '']
+          else:
+            continue
+          # make sure that all the single args have a boolean value
+          # associated so that we can load the args to a key value
+          # structure
+          args_list = config.insert_bool_values(value.split())
+          args_list_string = ' '.join(args_list)
+          if not  command or not app or not env:
+            Log.warn("heronrc config entry %s does not have key parameters (command:app:env) ",
+                     line)
+            continue
+          if app not in cls.cmdmap:
+            cls.cmdmap[app] = collections.defaultdict(dict)
+
+          if command in cls.cmdmap[app] and env in cls.cmdmap[app][command]:
+            cls.cmdmap[app][command][env] = cls.cmdmap[app][command][env] + ' ' + args_list_string
+          else:
+            cls.cmdmap[app][command][env] = args_list_string
+      Log.debug("RC cmdmap %s", json.dumps(cls.cmdmap))
+    else:
+      Log.warn("%s is not an existing file", effective_rc)
```

Comment:
@objmagic  - HeronRCArgumentParser extends Argparser, and hence the behavior will be defaulting to normal argparser behavior without the file. is that what you mean?